### ISSUE TITLE
Unicode not sourced properly

### ISSUE THREAD

**Issue by nOmi7** [processone/ejabberd#723]
I have configure mysql to store emojis per standards  http://apps.timwhitlock.info/emoji/tables/unicode
which it soruces properly
conntent of mysql field What a nice emoji😀!
however when i am running odbc:query() to fetch my mysql field it comes as 
"What a nice emoji?! " and it send the same over to the client.

How can i map http://apps.timwhitlock.info/emoji/tables/unicode to ejabberd  ? or i am doing something wrong

**Comment by nOmi7** [comment]
@badlop  any pointers on this ? its p1_mysql issue or something that i am doing wrong ?

**Comment by mremond** [comment]
Do you have your database configured properly with proper encoding ? You need to have a database with utf8mb4.
utf8 is not enough in MySQL setting to store emoji.

**Comment by prefiks** [comment]
Also i think you need to have at least mysql 5.3, as previous versions didn't handle unicode 6 that added emoji.

**Comment by nOmi7** [comment]
Yep i have mysql 5.5 configured with utf8mb4 but p1_mysql does not retrv it properly

**Comment by nOmi7** [comment]
@mremond  i have utf8mb4 when i run insert from mysql client it works as it should even for the selects
however ejabberd does some conversion list to binary where it messes up for sure 

```
    UPairs = lists:zipwith(fun (A, B) ->
				   <<A/binary, "='", B/binary, "'">>
			   end,
			   Fields, Vals),

```

emoji is not stored way it should  even if db is utf8mb4 
when i do select on manual insert is see it stored properly


MariaDB [ejabberd]> select  test_emoji from test where  1
+------------+
| test_emoji |
+------------+
| 🙈😄         |
+------------+
1 row in set (0.00 sec)

**Comment by prefiks** [comment]
Would you be able to see what `select hex(test_emoji) from test;` returns on that table for value from erlang?

You may also need execute `set names 'utf8mb4';` directly after connecting to be able to send/receive values in that extended utf8 format.

**Comment by elf0423** [comment]
My mate implement work around for this issue is change the DB column to blob

**Comment by ma2n** [comment]
i have same problem insert emoji, i set my table and column to utf8mb4_bin on database and character_client and server set to utf8mb4, but still got problem.....

Last abort reason: "#HY000Incorrect string value: '\\xF0\\x9F\\x98\\x98</...' for column 'vcard' at row 1".

**Comment by Nitin3588** [comment]
i would suggest if you are not able to do with mysql 5.5 for unicode (emoji)  support , you could go with  PostgreSQL . it works fine .

**Comment by cromain** [comment]
this was fully fixed by 7d1c75d0 and b2c6e39 in 16.02
